xds/internal/xdsclient/client_refcounted_test.go

```diff
@@ -0,0 +1,251 @@
+	// Create a bootstrap configuration, place it in a file in the temp
+	// directory, and set the bootstrap env vars to point to it.
+	nodeID := uuid.New().String()
+	contents := e2e.DefaultBootstrapContents(t, nodeID, "non-existent-server-address")
+	testutils.CreateBootstrapFileForTesting(t, contents)
+
+	// Override the client creation hook to get notified.
+	origClientImplCreateHook := xdsClientImplCreateHook
+	clientImplCreateCh := testutils.NewChannel()
+	xdsClientImplCreateHook = func(name string) {
+		clientImplCreateCh.Replace(name)
+	}
+	defer func() { xdsClientImplCreateHook = origClientImplCreateHook }()
+
+	// Override the client close hook to get notified.
+	origClientImplCloseHook := xdsClientImplCloseHook
+	clientImplCloseCh := testutils.NewChannel()
+	xdsClientImplCloseHook = func(name string) {
+		clientImplCloseCh.Replace(name)
+	}
+	defer func() { xdsClientImplCloseHook = origClientImplCloseHook }()
+
+	// Create two xDS clients.
+	client1Name := t.Name() + "-1"
+	_, closeFunc1, err := New(client1Name)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	name, err := clientImplCreateCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Timeout when waiting for xDS client to be created: %v", err)
+	}
+	if name.(string) != client1Name {
+		t.Fatalf("xDS client created for name %q, want %q", name.(string), client1Name)
+	}
+
+	client2Name := t.Name() + "-2"
+	_, closeFunc2, err := New(client2Name)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	name, err = clientImplCreateCh.Receive(ctx)
+	if err != nil {
+		t.Fatalf("Timeout when waiting for xDS client to be created: %v", err)
+	}
+	if name.(string) != client2Name {
+		t.Fatalf("xDS client created for name %q, want %q", name.(string), client1Name)
+	}
+
+	// Create N more references to each of these clients.
+	const count = 9
+	closeFuncs1 := make([]func(), count)
+	closeFuncs2 := make([]func(), count)
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+		for i := 0; i < count; i++ {
+			var err error
+			_, closeFuncs1[i], err = New(client1Name)
+			if err != nil {
+				t.Errorf("%d-th call to New() failed with error: %v", i, err)
+			}
+		}
+	}()
+	go func() {
+		defer wg.Done()
+		for i := 0; i < count; i++ {
+			var err error
+			_, closeFuncs2[i], err = New(client2Name)
+			if err != nil {
+				t.Errorf("%d-th call to New() failed with error: %v", i, err)
+			}
+		}
+	}()
+	wg.Wait()
+	if t.Failed() {
+		t.FailNow()
+	}
+
+	// Ensure that none of the create hooks are called.
+	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCancel()
+	if _, err := clientImplCreateCh.Receive(sCtx); err == nil {
+		t.Fatalf("New xDS client created when expected to reuse an existing one")
+	}
+
+	// Close the first client completely.
+	closeFunc1()
+	for i := 0; i < count; i++ {
+		closeFuncs1[i]()
```

Comment:
Optional: here and the close for 2: the test above tests the idempotency implemented through the once func. Maybe call some of these twice?

Reply:
Sure. Done.